src/Microsoft.AspNet.Buffering/ResponseBufferingMiddleware.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+
+namespace Microsoft.AspNet.Buffering
+{
+    public class ResponseBufferingMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public ResponseBufferingMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var originalBufferingFeature = httpContext.GetFeature<IHttpBufferingFeature>();
+            var originalSendFileFeature = httpContext.GetFeature<IHttpSendFileFeature>();
+            var originalResponseBody = httpContext.Response.Body;
+            try
+            {
+                // Shim the response stream
+                var bufferStream = new BufferingWriteStream(originalResponseBody);
+                httpContext.Response.Body = bufferStream;
+                httpContext.SetFeature<IHttpBufferingFeature>(new HttpBufferingFeature(bufferStream, originalBufferingFeature));
+                if (originalSendFileFeature != null)
+                {
+                    httpContext.SetFeature<IHttpSendFileFeature>(new SendFileFeatureWrapper(originalSendFileFeature, bufferStream));
+                }
+
+                await _next(httpContext);
+
+                // If we're still buffered, set the content-length header and flush the buffer.
+                // Only if the content-length header is not already set, and some content was buffered.
+                if (!httpContext.Response.HasStarted && bufferStream.CanSeek && bufferStream.Length > 0)
```

Comment:
if nothing has been written should we explicitly set the Content-Length to 0 (if it was not already set)?